### PR TITLE
All files updated to work with Hubspot API

### PR DESCRIPTION
**Pull request by BeAllAround**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,33 +30,33 @@ that they can be accessed using the Seneca entity API and messages.
 Seneca()
   .use('provider', {
     provider: {
-      trello: {
+      hubspot: {
```
Comment (rjrodger) [inline README.md:40]: update this to show usage with the `env` plugin as per test/live.js

File: README.md [hunk 2]
```diff
@@ -30,33 +30,33 @@ that they can be accessed using the Seneca entity API and messages.
...
       }
     }
   })
-  .use('trello-provider')
+  .use('hubspot-provider')
 
-let repo = await seneca.entity('provider/trello/repo')
-  .load$('senecajs/trello-api-test')
+let company = await seneca.entity('provider/hubspot/company')
+  .load$('id/type/name/city/state/description')
```
Comment (rjrodger) [inline README.md]: this should just be a company identifier

File: README.md [hunk 3]
```diff
@@ -73,7 +73,7 @@ Set plugin options when loading with:
 ```js
 
 
-seneca.use('TrelloProvider', { name: value, ... })
+seneca.use('HubspotProvider', { name: value, ... })
```
Comment (rjrodger) [inline README.md]: should be `hubspot-provider`

File: package.json [hunk 4]
```diff
@@ -1,20 +1,22 @@
...
+    "test": "test"
+  },
   "keywords": [
     "seneca",
-    "trello",
-    "trello-provider"
+    "hubspot",
+    "hubspot-provider"
   ],
   "author": "Richard Rodger - richardrodger.com",
```
Comment (rjrodger) [inline package.json:16]: Add yourself as a contributor :)
https://docs.npmjs.com/cli/v8/configuring-npm/package-json#people-fields-author-contributors

File: package.json [hunk 5]
```diff
@@ -1,20 +1,22 @@
...
   "repository": {
     "type": "git",
-    "url": "git://trello.com/senecajs/seneca-trello-provider.git"
+    "url": "git://hubspot.com/senecajs/seneca-hubspot-provider.git"
```
Comment (rjrodger) [inline package.json]: an error in the original - this should be github.com

File: package.json [hunk 6]
```diff
@@ -50,14 +52,16 @@
...
     "@seneca/doc": "^4.1.0",
     "@types/jest": "^28.1.3",
     "@types/node": "^18.0.0",
     "esbuild": "^0.14.47",
     "esbuild-jest": "^0.5.0",
     "jest": "^28.1.1",
-    "seneca": ">=3",
+    "seneca": "^3.28.2",
     "seneca-msg-test": "^3.4.0",
+    "ts-jest": "^28.0.5",
```
Comment (rjrodger) [inline package.json]: this should not be necessary as we use use esbuild for speed

File: src/hubspot-provider.d.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+declare type HubspotProviderOptions = {};
```
Comment (rjrodger) [inline src/hubspot-provider.d.ts]: this should not be needed as it is automatically created

File: src/hubspot-provider.ts [hunk 8]
```diff
@@ -0,0 +1,103 @@
+const Pkg = require('../package.json');
+const hubspot = require('@hubspot/api-client');
+
+type HubspotProviderOptions = {}
+
+function HubspotProvider(this: any, options : HubspotProviderOptions) {
+	const seneca:any = this;
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: seneca indent convention is 2 spaces - run prettier to get this
also we don't use semis

File: src/hubspot-provider.ts [hunk 9]
```diff
@@ -0,0 +1,103 @@
+const Pkg = require('../package.json');
+const hubspot = require('@hubspot/api-client');
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: capitalize - this is also a convention in Seneca code

File: src/hubspot-provider.ts [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+	entityBuilder(this, {
+		  provider: {
+			  name: 'hubspot'
+		  },
+		  entity: {
+			  company: {
+				  cmd: {
+					  list: {
+						  action: async function(this: any, entize: any, msg: any) {
+							  let res, list;
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: unless you are specifying types, no need to predefine vars

File: src/hubspot-provider.ts [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+							  res = (await this.shared.sdk.crm.companies.basicApi.getPage(10)).results;
```
Comment (rjrodger) [inline src/hubspot-provider.ts]:
```
const res = ...
let list = ...
```

avoid hard-coded values in code: `getPage(10)`
this value should come from `msg.q.limit$` see https://senecajs.org/docs/tutorials/understanding-query-syntax.html#wp-limit

File: src/hubspot-provider.ts [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+							  list = res.map((data) => entize(data))
+							  return list;
+						  }
+					  },
+					  load: {
+						  action: async function(this: any, entize: any, msg: any) {
+							  let data = (msg.q.id.split('/'));
+							  let id = data[0];
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: what does hubspot use as the unique id from companies? that should be sufficient

File: src/hubspot-provider.ts [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+							  let obj = {};
+							  data = data.slice(1);
+							  try{
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: no need for try catch here - the function is async so the exception will be caught further up the call stack

File: src/hubspot-provider.ts [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+						  action: async function(this: any, entize: any, msg: any) {
+							  let res, list;
+							  res = (await this.shared.sdk.crm.companies.basicApi.getPage(10)).results;
+							  list = res.map((data) => entize(data))
+							  return list;
+						  }
+					  },
+					  load: {
+						  action: async function(this: any, entize: any, msg: any) {
+							  let data = (msg.q.id.split('/'));
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: unless the value really is generic "data" avoid using `data` as a variable name
in this case `idParts` might be better

File: src/hubspot-provider.ts [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+							  let id = data[0];
+							  let obj = {};
+							  data = data.slice(1);
+							  try{
+								  obj = await this.shared.sdk.crm.companies.basicApi.getById(id, data.length != 0 ? data : undefined);
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: what a strange API! - please add a comment with a link to the hubspot docs for this method

File: test/basic.messages.js [hunk 16]
```diff
@@ -1,22 +1,20 @@
...
       pattern: 'get:info',
       out: {
         ok: true,
-        name: 'trello',
+        name: 'hubspot',
         version: Pkg.version,
         sdk: {
-          name: 'trello-node',
-          version: Pkg.dependencies['trello'],
+          name: 'hubspot-node',
```
Comment (rjrodger) [inline test/basic.messages.js]: the package name is just `hubspot`

**Review by rjrodger — CHANGES_REQUESTED** [review]

**Comment by BeAllAround** [comment]
I committed all of your changes!

Regards!

File: README.md (hunk 2 above)
Comment (rjrodger) [reply README.md]: TODO @rjrodger explain Seneca fields$ directive

File: src/hubspot-provider.ts (hunk 8 above)
Comment (BeAllAround) [reply src/hubspot-provider.ts]: Done!

File: README.md (hunk 2 above)
Comment (rjrodger) [reply README.md]: @BeAllAround see https://github.com/senecajs/seneca-entity/blob/master/test/entity.test.js#L377 for the "seneca way" to do this

the .load$ method should only accept either a single value as id, or a query object.

return all the standard hubspot fields by default, or parse the fields$ directive to get the fields that the user wants

```
seneca.entity('provider/hubspot/company').load$('ID-VALUE')
seneca.entity('provider/hubspot/company').load$({id:'ID-VALUE'})
seneca.entity('provider/hubspot/company').load$({someField:'FIELD-VALUE'}) // same as SQL WHERE someField = "FIELD-VALUE"

seneca.entity('provider/hubspot/company').load$({id:'ID-VALUE', fields$:['type','name']}) // only load the type and name fields
```

File: src/HubspotProvider-doc.d.ts [hunk 17]
```diff
@@ -0,0 +1,6 @@
+declare const docs: {
```
Comment (rjrodger) [inline src/HubspotProvider-doc.d.ts]: shouldn't this be in the dist folder?

File: src/HubspotProvider-doc.ts [hunk 18]
```diff
@@ -0,0 +1,12 @@
+const docs = {
+	get_info: {
```
Comment (rjrodger) [inline src/HubspotProvider-doc.ts]: indentation should be 2 spaces

File: src/hubspot-provider.d.ts [hunk 19]
```diff
@@ -0,0 +1,6 @@
+declare function HubspotProvider(this: any, options: HubspotProviderOptions): {
```
Comment (rjrodger) [inline src/hubspot-provider.d.ts]: also, should this not be in dist?
or did you define this manually?

File: src/hubspot-provider.ts [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+type HubspotProviderOptions = {}
+
+function HubspotProvider(this: any, options : HubspotProviderOptions) {
+  const seneca:any = this
+
+  const entityBuilder = this.export('provider/entityBuilder')
+
+  seneca.message('sys:provider,provider:hubspot,get:info', get_info)
+  async function get_info(this: any, _msg: any) {
```
Comment (rjrodger) [inline src/hubspot-provider.ts:15]: add a line to separate here
if more messages are added later, we want all the messages to be together
see: https://github.com/mgp/book-notes/blob/master/the-non-designers-design-book.markdown - proximity

File: src/hubspot-provider.ts [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+      ok: true,
+      name: 'hubspot',
+      version: Pkg.version,
+      sdk: {
+        name: 'hubspot',
+	version: Pkg.dependencies['hubspot'],
+      }
+    }
+  }
+ entityBuilder(this, {
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: same here - line separation needed

Comment (rjrodger) [reply src/hubspot-provider.ts]: also, the indentation is incorrect

File: src/hubspot-provider.ts [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+   provider: {
+     name: 'hubspot'
+   },
+   entity: {
+     company: {
+       cmd: {
+         list: {
+	     action: async function(this: any, entize: any, msg: any) {
+	       const limit = msg.q.limit || 10 // The maximum number of results to display per page
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: use `limit$`

File: src/hubspot-provider.ts [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+	       let list = res.map((data) => entize(data))
+	       return list
+	     }
+	 },
+
+	 load: {
+	   action: async function(this: any, entize: any, msg: any) {
+	     let idParts = msg.q.fields$
+	     let id = msg.q.id
+	     let obj = await this.shared.sdk.crm.companies.basicApi.getById(id, idParts) // docs for the usage: https://developers.hubspot.com/docs/api/crm/companies
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: this line is too long - aim for max 80 chars (makes things like diff comparison easier)
place the comment above the line

File: src/hubspot-provider.ts [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+	   }
+	 },
+
+	 save: {
+	   action: async function(this: any, entize: any, msg: any) {
+	     let ent = msg.ent
+	     let id = ent.id
+	     let desc = ent.properties.description
+	     let obj
+	     try{
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: missing a space - just run https://prettier.io and all these formatting problems go away!

File: src/hubspot-provider.ts [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+	       throw err
+	     }
+	     return entize(obj)
+	   }
+	 }
+       }
+     }
+   }
+ })
+
```
Comment (rjrodger) [inline src/hubspot-provider.ts]: too much space

in general, use 2 lines for "major" contextual differences

**Review by rjrodger — CHANGES_REQUESTED** [review]

**Comment by BeAllAround** [comment]
> @@ -0,0 +1,6 @@
+declare function HubspotProvider(this: any, options: HubspotProviderOptions): {

also, should this not be in dist?
or did you define this manually?


This was in the Trello provider like that.

**Comment by BeAllAround** [comment]
You are right actually. Sorry about that. It seems like jest generated it

**Comment by BeAllAround** [comment]
I think we are all set now!

**Pull request merged**